src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -249,179 +256,149 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
-                                    {
-                                        refreshAll = true;
-                                        break;
-                                    }
-                                }
-                                else
-                                {
-                                    logDebugBuilder.AppendLine(LogHelper.BuildKeyValueReadMessage(change.ChangeType, change.Key, change.Label, endpoint.ToString()));
-                                }
+                                refreshAll = await HaveCollectionsChanged(
+                                    _options.Selectors.Where(selector => !selector.IsFeatureFlagSelector).ToList(),
+                                    _kvEtags,
+                                    client,
+                                    cancellationToken).ConfigureAwait(false);
                             }
+                        }
+                        else
+                        {
+                            refreshAll = await RefreshIndividualKvWatchers(
+                                client,
+                                keyValueChanges,
+                                refreshableIndividualKvWatchers,
+                                endpoint,
+                                logDebugBuilder,
+                                logInfoBuilder,
+                                cancellationToken).ConfigureAwait(false);
+                        }
 
-                            if (refreshAll)
+                        if (refreshAll)
+                        {
+                            // Trigger a single load-all operation if a change was detected in one or more key-values with refreshAll: true
+                            kvEtags = new Dictionary<KeyValueSelector, IEnumerable<MatchConditions>>();
+                            ffEtags = new Dictionary<KeyValueSelector, IEnumerable<MatchConditions>>();
+
+                            data = await LoadSelected(client, kvEtags, ffEtags, _options.Selectors, cancellationToken).ConfigureAwait(false);
+                            watchedIndividualKvs = await LoadKeyValuesRegisteredForRefresh(client, data, cancellationToken).ConfigureAwait(false);
+                            logInfoBuilder.AppendLine(LogHelper.BuildConfigurationUpdatedMessage());
+                            return;
+                        }
+                        else
+                        {
+                            // Get feature flag changes
+                            ffCollectionUpdated = await HaveCollectionsChanged(
+                                refreshableFfWatchers.Select(watcher => new KeyValueSelector { KeyFilter = watcher.Key, LabelFilter = watcher.Label }),
+                                _ffEtags,
+                                client,
+                                cancellationToken).ConfigureAwait(false);
+
+                            if (ffCollectionUpdated)
                             {
```

Comment:
This cloned dictionary doesn't seem necessary given HaveCollectionsChanged doesn't modify input.

Reply:
I think the changes may not have loaded yet, removed in previous commit